### DISCUSSION TITLE
An issue querying DuckDB has a vector datastore...

### DISCUSSION THREAD

**Discussion by charnould** [duckdb/duckdb#12432] [Q&A] [answered]
First, congrats for DuckDB 1.0!

I'm toying with DuckDB using it as a datastore enabling full text search + vector search (I'm building a small AI app).
Full text search works fine/perfectly, but I'm struggling with vector search.

Somewhere, I've created this database:  
(`embedding` is generated using `text-embedding-3-small` from OpenAI)

`con.sql("CREATE TABLE knowledge (id VARCHAR, content VARCHAR, embedding DECIMAL[1536]);");
`

Somewhere else, I generate embeddings for a query using the very same `text-embedding-3-small` .

```
const embedding = [ -0.0037245885, 0.040825516, -0.013087831, -0.0061711613, -0.050031964, -0.009673872,
  -0.046836924, -0.026104663, -0.026412332, -0.0098927915, 0.025205318, -0.05201999,
  0.039121494, -0.03673113, -0.055948704, 0.039168827, 0.04295554, 0.022329781, -0.006910754
....truncated........ ]
```

and then I try to query DuckDB:

```
con.all(`
            SELECT *
            FROM (
            SELECT *, array_cosine_similarity(embedding, [${embedding}]) AS score
            FROM knowledge
            ) sq
            WHERE score IS NOT NULL
            ORDER BY score DESC LIMIT 10;`,
			(err, res) => {
				if (err) {
					console.warn(err);
					return;
				}

				console.log("EMBED:" , res);
			},
		);
```

But I've got an error:

```
error: Binder Error: No function matches the given name and argument types 'array_cosine_similarity(DECIMAL(18,3)[1536], DECIMAL(13,12)[])'. You might need to add explicit type casts.
        Candidate functions:
        array_cosine_similarity(FLOAT[ANY], FLOAT[ANY]) -> FLOAT
        array_cosine_similarity(DOUBLE[ANY], DOUBLE[ANY]) -> DOUBLE

LINE 4:             SELECT *...
```

Error message is clear, but I cannot find how to fix it.
1. I do not understand why I must wrap my embeddings in `[]`
2. I do not understand also why it says `DECIMAL(13,12)`when I do not have 12 decimal digits.
3. Why `DECIMAL(18,3)[1536]` and `DECIMAL(13,12)[]`?

Bref, any clue how to fix this?
Thanks a lot.

-- EDIT --

When I create a database using `DOUBLE`...

`con.sql("CREATE TABLE knowledge (id VARCHAR, content VARCHAR, embedding DOUBLE[1536]);");`

error is...

```
error: Binder Error: No function matches the given name and argument types 'array_cosine_similarity(DOUBLE[1536], DECIMAL(13,12)[])'. You might need to add explicit type casts.
        Candidate functions:
        array_cosine_similarity(FLOAT[ANY], FLOAT[ANY]) -> FLOAT
        array_cosine_similarity(DOUBLE[ANY], DOUBLE[ANY]) -> DOUBLE

LINE 4:             SELECT *...
```

**Comment by Tishj** [comment]
I think the issue is that you're creating a DECIMAL column instead of FLOAT/DOUBLE, there is no overload for DECIMAL for that function

**Reply by charnould** [reply]
Thanks. I've edited my question, but the result is the very same....

```
error: Binder Error: No function matches the given name and argument types 'array_cosine_similarity(DOUBLE[1536], DECIMAL(13,12)[])'. You might need to add explicit type casts.
        Candidate functions:
        array_cosine_similarity(FLOAT[ANY], FLOAT[ANY]) -> FLOAT
        array_cosine_similarity(DOUBLE[ANY], DOUBLE[ANY]) -> DOUBLE

LINE 4:             SELECT *...
```

**Reply by Tishj** [reply]
Might need to add a cast to `[${embedding}]`
`[${embedding}]::DOUBLE[]`

**Reply by charnould** [reply, marked as answer]
`[${embedding}]::DOUBLE[1536]` works like a charm! Thanks a lot.
